File: src/main/java/tooko/td/client/TdBot.java
package tooko.td.client;

import cn.hutool.core.io.FileUtil;
import cn.hutool.core.text.csv.CsvData;
import cn.hutool.core.text.csv.CsvRow;
import cn.hutool.core.text.csv.CsvUtil;
import cn.hutool.core.text.csv.CsvWriter;
import cn.hutool.core.thread.ThreadUtil;
import cn.hutool.core.util.*;
import tooko.main.Env;
import tooko.main.Fn;
import tooko.main.Lang;
import tooko.td.TdApi;
import tooko.td.TdApi.*;
import tooko.td.core.LongLongArrayMap;

import java.io.File;
import java.util.HashMap;
import java.util.*;

public class TdBot extends TdClient {

    public File dataDir;
    public LongLongArrayMap startMessages = new LongLongArrayMap();
    public HashMap<String, LongLongArrayMap> startPayloads = new HashMap<>();
    public HashMap<Integer, TdPersistent> persists = new HashMap<>();
    public HashMap<Integer, TdHandler> persistHandlers = new HashMap<>();
    public HashMap<String, TdHandler> payloads = new HashMap<>();
    public HashMap<String, TdHandler> functions = new HashMap<>();
    public HashMap<Integer, TdHandler> callbacks = new HashMap<>();
    public String botToken;

    public TdBot(String botToken) {

        this(botToken, new TdOptions().databaseDirectory(initDataDir(botToken)));

        this.dataDir = Env.getFile("data/" + StrUtil.subBefore(botToken, ":", false));

    }

    public TdBot(String botToken, TdOptions options) {

        super(options);

        this.botToken = botToken;

    }

    private static String initDataDir(String botToken) {

        File dataDir = Env.getFile("data/" + StrUtil.subBefore(botToken, ":", false));

        dataDir.mkdirs();

        mkLink(dataDir, "stickers");
        mkLink(dataDir, "profile_photos");
        mkLink(dataDir, "thumbnails");
        mkLink(dataDir, "wallpapers");

        return dataDir.getPath();

    }

    private static void mkLink(File dataDir, String target) {

        File sourceDir = new File(dataDir, target);

        File targetDir = Env.getFile("cache/files/" + target);

        if (!sourceDir.isDirectory()) {

            targetDir.mkdirs();

            RuntimeUtil.execForStr("ln -s " + targetDir.getPath() + " " + sourceDir.getPath());

        }

    }

    @Override
    public void onAuthorizationState(AuthorizationState authorizationState) {

        super.onAuthorizationState(authorizationState);

        if (authorizationState instanceof TdApi.AuthorizationStateWaitPhoneNumber) {

            try {

                execute(new TdApi.CheckAuthenticationBotToken(botToken));

            } catch (TdException ex) {

                onAuthorizationFailed(ex);

            }

        }
    }

    public void onAuthorizationFailed(TdException ex) {
    }

 

    @Override
    public void onLoad() {

        File file = new File(dataDir, "persist.csv");

        if (file.isFile()) {

            CsvData csv = CsvUtil.getReader().read(file);

            for (CsvRow row : csv.getRows()) {

                int userId = NumberUtil.parseInt(row.get(0));
                int dataId = NumberUtil.parseInt(row.get(1));
                int subId = NumberUtil.parseInt(row.get(2));
                long createAt = NumberUtil.parseLong(row.get(3));
                boolean acceptFunction = NumberUtil.parseInt(row.get(4)) == 1;

                String[] data = Fn.shift(row.getRawList(), 4).toArray(new String[row.getFieldCount() - 4]);

                if (persistHandlers.containsKey(dataId)) {

                    persists.put(userId, new TdPersistent(userId, dataId, subId, createAt, acceptFunction));

                    persistHandlers.get(dataId).onPersistRestore(userId, subId, data);

                }

            }

            FileUtil.del(file);

        }

    }

    @Override
    public void onDestroy() {

        if (!persists.isEmpty()) {

            File file = new File(dataDir, "persist.csv");

            CsvWriter csv = CsvUtil.getWriter(file, CharsetUtil.CHARSET_UTF_8, false);

            for (TdPersistent persist : persists.values()) {

                if ((System.currentTimeMillis() - persist.createAt) > 3 * 60 * 60 * 1000L || persistHandlers.get(persist.dataId).rejectPersistStore(persist.id, persist.subId)) {

                    persistHandlers.get(persist.dataId).onPersistTimeouted(persist.id, persist.subId);

                    continue;

                }

                LinkedList<String> data = new LinkedList<>();

                Collections.addAll(data,new String[]{persist.id + "", persist.dataId + "", persist.subId + "", persist.createAt + "", (persist.acceptFunction ? 1 : 0) + ""});
                
                persistHandlers.get(persist.dataId).onPersistStore(persist.id, persist.subId, data);
                
                csv.write(Fn.toArray(data,String.class));
                
            }

            csv.flush();

            csv.close();

        }

    }

    @Override
    public void onNewMessage(int userId, long chatId, TdApi.Message message) {

        while (!hasAuthed()) ThreadUtil.sleep(233);

        if (userId == me.id) return;

        User user = E(new GetUser(userId));

        if (Fn.fromPrivate(message)) {

            synchronized (persists) {

                if (persists.containsKey(userId)) {

                    TdPersistent persist = persists.get(userId);

                    persistHandlers.get(persist.dataId).handlePersistMessage(user, chatId, message, persist.subId, persist.acceptFunction);

                    Fn.finishEvent();

                }

            }

        }

        if (!(message.content instanceof MessageText)) return;

        FormattedText content = ((MessageText) message.content).text;

        if (content.entities.length == 0 || !(content.entities[0].type instanceof TextEntityTypeBotCommand) || content.entities[0].offset != 0) {

            return;

        }

        TextEntity command = content.entities[0];

        String function = content.text.substring(1, command.length);

        if (function.endsWith("@" + me.username)) {

            function = function.substring(0, function.length() - me.username.length() - 1);

        }

        // log.debug("{} : FN /{}",Fn.displayName(user),function);

        String param = content.text.length() > command.length ? content.text.substring(command.length + 1) : null;

        String[] params;
        String[] originParams;

        if (param == null) {

            param = "";
            params = originParams = new String[0];

        } else {

            originParams = param.split(" ");
            params = param.replace("  ", " ").split(" ");

        }

        try {

            if ("start".equals(function)) {

                if (StrUtil.isBlank(param)) {

                    onLaunch(user, message.chatId, message);

                } else {

                    String data = params[0];

                    String prefix = StrUtil.subBefore(data, "_", false);

                    params = StrUtil.subAfter(data, "_", false).split("_");

                    if (!payloads.containsKey(prefix)) {

                        onUnDefPayload(user, chatId, message, prefix, params);

                        Fn.finishEvent();

                    }

                    payloads.get(prefix).onPayload(user, chatId, message, prefix, params);


                }

                Fn.finishEvent();

            }

            if (!functions.containsKey(function)) {

                onUnDefFn(user, message.chatId, message, function, param, params, originParams);

            } else {

                functions.get(function).onFunction(user, message.chatId, message, function, param, params, originParams);

            }

        } catch (Reject reject) {

            return;

        } catch (TdException ex) {

            if (ex.getCode() == -1) {

                return;

            }

            throw ex;

        }

        Fn.finishEvent();

    }

    @Override
    public void handleNewCallbackQuery(long queryId, int senderUserId, long chatId, long messageId, long chatInstance, TdApi.CallbackQueryPayload payload) {

        if (payload instanceof CallbackQueryPayloadGame) return;

        byte[] data = ((CallbackQueryPayloadData) payload).data;

        if (data[0] == 120 && data[1] == -38) {

            data = ZipUtil.unZlib(data);

        }

        int id = data[0] + 129;

        int subId = data[1];

        if (!callbacks.containsKey(id)) {

            send(Fn.answerAlert(queryId, 114, "Invalid Data #{}", id));

            return;

        }

        try {

            execute(new GetMessage(chatId, messageId));

        } catch (TdException e) {

            if (e.getCode() == -1) {

                send(Fn.answerAlert(queryId, Lang.get(senderUserId).RESTARTING_WARN));

            } else {

                send(Fn.answerAlert(queryId, 114, "Invalid Query : {}", e));

            }

            return;

        }

        callbacks.get(id).onNewCallbackQuery(senderUserId, chatId, messageId, queryId, subId, Fn.readData(data));

        Fn.finishEvent();

    }


    public void onLaunch(User user, long chatId, Message message) {

        deleteStartMessages(chatId);

        send(Fn.sendSticker(chatId, "CAADAgADJAMAAsEYngsXfXVKAnZzKxYE"), asStartMessage(message));

    }

    public void onUnDefPayload(User user, long chatId, Message message, String payload, String[] params) {

        if (Fn.fromPrivate(message)) {

            send(Fn.deleteMessage(message));

        }

    }

    public void onUnDefFn(User user, long chatId, Message message, String function, String param, String[] params, String[] originParams) {

        if (Fn.fromPrivate(message)) {

            Lang L = Lang.get(user);

            delay(Fn.sendText(chatId, Fn.plainText(L.CNF)));

            delay(message);

        }

    }

}

File: src/main/java/tookox/core/client/TdBot.kt
package tookox.core.client

import cn.hutool.core.util.RuntimeUtil
import cn.hutool.core.util.StrUtil
import cn.hutool.core.util.ZipUtil
import kotlinx.coroutines.coroutineScope
import tooko.main.Env
import tooko.main.Fn
import tooko.td.TdApi.*
import tookox.core.finishEvent
import tookox.core.fromPrivate
import tookox.core.shift
import tookox.core.utils.makeAnswer
import tookox.core.utils.readDataFrom
import tookox.core.utils.writeDataTo
import java.io.File
import java.util.*

open class TdBot(val botToken: String) : TdClient(initDataDir(botToken)), TdBotAbsHandler {

    val botUserId = botToken.substringBefore(':').toInt()

    override val sudo: TdBot get() = this

    val payloads = HashMap<String, TdBotAbsHandler>()
    val functions = HashMap<String, TdBotAbsHandler>()
    val callbacks = HashMap<Int, TdBotAbsHandler>()

    val persistHandlers = HashMap<Int, TdBotAbsHandler>()
    val persists = HashMap<Int, TdPerstst>()

    override suspend fun onLogin() {

        readDataFrom("user_persists")?.forEach {

            val userId = it[0].toInt()

            val persistId = it[1].toInt()

            val subId = it[2].toInt()

            val allowFuction = it[3].toInt() == 1

            val allowCancel = it[4].toInt() == 1

            val createAt = it[5].toInt()

            val data = it.shift(5).toList()

            val handler = (persistHandlers[persistId] ?: error("Invalid Persist ID #${persistId}"))

            runCatching {

                handler.onPersistReStore(userId, subId, data)

            }.onFailure {

                return@forEach

            }

            sudo.persists[userId] = TdPerstst(userId, persistId, subId, allowFuction, allowCancel, createAt)

        }

    }

    override suspend fun onDestroy() {

        super<TdBotAbsHandler>.onDestroy()

        val dataList = LinkedList<List<String>>()

        persists.forEach { (userId, persist) ->

            val data = LinkedList<String>()

            data.add("$userId")

            data.add("${persist.persistId}")

            data.add("${persist.subId}")

            data.add("${persist.allowFuction}")

            data.add("${persist.allowCancel}")

            data.add("${persist.createAt}")

            data.add("${persist.persistId}")

            runCatching {

                onPersistStore(userId, persist.subId, data)

            }.onFailure {

                return@forEach

            }

            dataList.add(data)

        }

        writeDataTo("user_persists", dataList)

    }

    override suspend fun onAuthorizationState(authorizationState: AuthorizationState) = coroutineScope {

        super.onAuthorizationState(authorizationState)

        if (authorizationState is AuthorizationStateWaitPhoneNumber) {

            sendUnit(CheckAuthenticationBotToken(botToken)) onError {

                onAuthorizationFailed(it)

            }

        }

    }

    override suspend fun onNewMessage(userId: Int, chatId: Long, message: Message) = coroutineScope {

        if (auth && userId == me.id) return@coroutineScope

        val persist = if (message.fromPrivate && persists.containsKey(userId)) {

            persists[userId]

        } else null

        run predict@{

            if (message.content !is MessageText) return@predict

            val content = (message.content as MessageText).text

            var param = content.text

            run fn@{

                Env.FUN_PREFIX.forEach {

                    if (!param.startsWith(it)) return@forEach

                    param = param.substring(it.length)

                    return@fn

                }

                return@predict

            }

            var function = if (!param.contains(' ')) {

                param.also {

                    param = ""

                }

            } else {

                param.substringBefore(' ').also {

                    param = param.substringAfter(' ')

                }

            }

            val validSuffix = "@${me.username}"

            if (function.endsWith(validSuffix)) {

                function = function.substring(0, function.length - validSuffix.length)

            }

            val params: Array<String>

            val originParams: Array<String>

            if (param.isBlank()) {

                originParams = arrayOf()
                params = originParams

            } else {

                originParams = param.split(' ').toTypedArray()
                params = param.replace("  ", " ").split(' ').toTypedArray()

            }

            if (persist != null) run persist@{

                val handler = persistHandlers[persist.persistId] ?: error("Invalid Persist ID #${persist.persistId}")

                if (function == "cancel" && persist.allowCancel) {

                    sudo removePersist userId

                    handler.onPersistCancel(userId, chatId, message, persist.subId)

                    handler.onPersistRemoveOrCancel(userId, persist.subId)

                    handler.onSendCanceledMessage(userId)

                    return@coroutineScope

                }

                if (persist.allowFuction) {

                    sudo removePersist userId

                    handler.onPersistCancel(userId, chatId, message, persist.subId)

                    handler.onPersistRemoveOrCancel(userId, persist.subId)

                    handler.onSendCanceledMessage(userId)

                    return@persist

                }

                handler.onPersistFunction(userId, chatId, message, persist.subId, function, param, params, originParams)

                return@coroutineScope

            }

            if ("start" == function) {

                if (params.isEmpty()) {

                    onLaunch(userId, chatId, message)

                } else {

                    // TODO: handle start payload

                }

            } else if (!functions.containsKey(function)) {

                handlers.filterIsInstance<TdBotAbsHandler>().forEach {

                    it.onUndefinedFunction(userId, chatId, message, function, param, params, originParams)

                }

            } else {

                functions[function]!!.onFunction(userId, chatId, message, function, param, params, originParams)

            }

            return@coroutineScope

        }

        if (persist == null) return@coroutineScope

        val handler = (persistHandlers[persist.persistId] ?: error("Invalid Persist ID #${persist.persistId}"))

        handler.onPersistMessage(userId, chatId, message, persist.subId)

        finishEvent()

    }

    override suspend fun handleNewInlineCallbackQuery(id: Long, senderUserId: Int, inlineMessageId: String, chatInstance: Long, payload: CallbackQueryPayload) {

        if (payload is CallbackQueryPayloadGame) return

        var data = (payload as CallbackQueryPayloadData).data

        if (data[0].toInt() == 120 && data[1].toInt() == -38) {

            data = ZipUtil.unZlib(data)

        }

        val dataId = data[0] + 129

        val subId = data[1].toInt()

        if (!callbacks.containsKey(dataId)) {

            sudo makeAnswer "Invalid Data #$id" answerTo id

            return

        }

        callbacks[dataId]!!.onNewInlineCallbackQuery(senderUserId, inlineMessageId, id, subId, Fn.readData(data))

        finishEvent()

    }

    override suspend fun handleNewCallbackQuery(id: Long, senderUserId: Int, chatId: Long, messageId: Long, chatInstance: Long, payload: CallbackQueryPayload) {

        if (payload is CallbackQueryPayloadGame) return

        var data = (payload as CallbackQueryPayloadData).data

        if (data[0].toInt() == 120 && data[1].toInt() == -38) {

            data = ZipUtil.unZlib(data)

        }

        val dataId = data[0] + 129

        val subId = data[1].toInt()

        if (!callbacks.containsKey(dataId)) {

            sudo makeAnswer "Invalid Data #$id" answerTo id

            return

        }

        callbacks[dataId]!!.onNewCallbackQuery(senderUserId, chatId, messageId, id, subId, Fn.readData(data))

        finishEvent()

    }

    open fun onLaunch(userId: Int, chatId: Long, message: Message) = Unit

    override suspend fun onFunction(userId: Int, chatId: Long, message: Message, function: String, param: String, params: Array<String>, originParams: Array<String>) = Unit
    override suspend fun onUndefinedFunction(userId: Int, chatId: Long, message: Message, function: String, param: String, params: Array<String>, originParams: Array<String>) = Unit
    override suspend fun onNewCallbackQuery(userId: Int, chatId: Long, messageId: Long, queryId: Long, subId: Int, data: Array<ByteArray>) = Unit
    override suspend fun onNewInlineCallbackQuery(userId: Int, inlineMessageId: String, queryId: Long, subId: Int, data: Array<ByteArray>) = Unit

    override suspend fun onPersistMessage(userId: Int, chatId: Long, message: Message, subId: Int) = Unit
    override suspend fun onPersistRemove(userId: Int, subId: Int) = Unit
    override suspend fun onPersistTimeout(userId: Int, subId: Int) = Unit
    override suspend fun onPersistRemoveOrCancel(userId: Int, subId: Int) = Unit
    override fun onPersistStore(userId: Int, subId: Int, data: LinkedList<String>) = Unit
    override fun onPersistReStore(userId: Int, subId: Int, data: List<String>) = Unit

    companion object {

        private fun initDataDir(botToken: String): TdOptions {

            val dataDir = Env.getFile("data/" + StrUtil.subBefore(botToken, ":", false))

            dataDir.mkdirs()

            mkLink(dataDir, "stickers")
            mkLink(dataDir, "profile_photos")
            mkLink(dataDir, "thumbnails")
            mkLink(dataDir, "wallpapers")


            return TdOptions().databaseDirectory(dataDir.path)

        }

        private fun mkLink(dataDir: File, target: String) {

            val sourceDir = File(dataDir, target)

            val targetDir = Env.getFile("cache/files/$target")

            if (!sourceDir.isDirectory) {

                targetDir.mkdirs()

                RuntimeUtil.execForStr("ln -s " + targetDir.path + " " + sourceDir.path)

            }

        }

    }

}